### PR TITLE
fix: Wait for install to be reported in state tree

### PR DESCRIPTION
**Pull request by adamspofford-dfinity**
Fixes SDK-544.

No way of e2e testing this for now unfortunately. It successfully installed a canister on mainnet without a spurious error, but to see if it actually fixed the problem will require waiting for it to flake and seeing that it doesn't, without replica throttling functionality.

### CODE REVIEW

File: src/dfx/src/lib/operations/canister/install_canister.rs [hunk 1]
```diff
@@ -136,7 +140,47 @@ pub async fn install_canister(
...
+                            env.get_logger(),
+                            "Waiting for module change to be reflected in system state tree..."
+                        )
+                    }
+                    sleep(Duration::from_millis(500)).await;
+                } else {
+                    bail!("The reported module hash is neither the existing module or the new one. \
+                        It has likely been modified while this command is running. \
+                        The state of the canister is unknown. \
+                        For this reason, no post-installation tasks have been run, including asset uploads.")
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/operations/canister/install_canister.rs:168]: Can we display the three module hashed involved here?  That will make it easier for the developer to follow up with `dfx canister info` and troubleshoot.

File: src/dfx/src/lib/operations/canister/install_canister.rs [hunk 2]
```diff
@@ -136,7 +140,47 @@ pub async fn install_canister(
...
+            }
+            Err(AgentError::LookupPathAbsent(_) | AgentError::LookupPathUnknown(_)) => {
+                times += 1;
+                if times % 4 == 0 {
+                    info!(
+                        env.get_logger(),
+                        "Waiting for module change to be reflected in system state tree..."
+                    )
+                }
+                sleep(Duration::from_millis(500)).await;
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/operations/canister/install_canister.rs:179]: What would you think about using a Waiter here, with some finite timeout?

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: src/dfx/src/lib/operations/canister/install_canister.rs [hunk 3]
```diff
@@ -113,14 +114,17 @@ pub async fn install_canister(
...
-        && wasm_module_already_installed(&wasm_module, installed_module_hash.as_deref())
+        && installed_module_hash
+            .as_deref()
+            .map_or(true, |old_hash| old_hash == new_hash)
         && !upgrade_unchanged
     {
         println!(
             "Module hash {} is already installed.",
-            hex::encode(installed_module_hash.unwrap())
+            hex::encode(installed_module_hash.as_ref().unwrap())
```
[the remark this answers is not in the document]
Comment (adamspofford-dfinity) [reply src/dfx/src/lib/operations/canister/install_canister.rs:127]: You're right, I got the condition mixed up.

File: src/dfx/src/lib/operations/canister/install_canister.rs (hunk 2 above)
Comment (adamspofford-dfinity) [reply src/dfx/src/lib/operations/canister/install_canister.rs:181]: I'm not really sure how Waiter works, but it looks blocking, which doesn't play well with async.

Comment (adamspofford-dfinity) [reply src/dfx/src/lib/operations/canister/install_canister.rs:181]: Wait, is that how we managed to starve the single-threaded web server however long ago?

Comment (ericswanson-dfinity) [reply src/dfx/src/lib/operations/canister/install_canister.rs:181]: That's why we use Garcon, which provides async_await().  Example:

https://github.com/dfinity/sdk/blob/master/src/dfx/src/commands/ledger/mod.rs#L165

Comment (adamspofford-dfinity) [reply src/dfx/src/lib/operations/canister/install_canister.rs:181]: Hmm. Some of those just block the thread anyway and then return a ready future, and others spin a new thread and sleep it. Both of these things defeat all benefits of asynchrony.

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]
Please add an entry to the changelog

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**